### PR TITLE
Add new sql module

### PR DESCRIPTION
**Pull request by dai-chen**
*Issue #, if available:* https://github.com/opendistro-for-elasticsearch/sql/issues/264, https://github.com/opendistro-for-elasticsearch/sql/issues/256

*Description of changes:* Add new SQL module to integrate with new query engine. As our start point, `SELECT` literal support is added. 

Here are the major changes:

 1. Add a new `sql` Gradle module.
 2. Support `SELECT` literal by:
  2.1 Relaxing existing `Project` operator to accept `Expression` instead of `ReferenceExpression`.
  2.2 Adding a new `Values` operator for literal values which is correspondent with ANSI `VALUES` clause. Reference: https://modern-sql.com/feature/values, https://crate.io/docs/crate/reference/en/latest/sql/statements/values.html.
 3. Integrate with legacy REST handler to route request to new SQL frontend.

The `legacy` module with all old code will co-exist with new `sql` module during code migration.

image

At runtime, request will send to new SQL parser first and fall back to old code if not supported yet.

image

*Documentation*: https://github.com/dai-chen/sql/blob/add-new-sql-module/docs/user/dql/expressions.rst

*Testing*: Passed old and new UT, IT and doctest.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: legacy/src/main/java/com/amazon/opendistroforelasticsearch/sql/legacy/plugin/RestSqlAction.java [hunk 1]
```diff
@@ -135,8 +142,19 @@ protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient cli
...
+            Format format = SqlRequestParam.getFormat(request.params());
+
+            // Route request to new query engine if it's supported already
+            SQLQueryRequest newSqlRequest = new SQLQueryRequest(sqlRequest.getJsonContent(),
+                                                                sqlRequest.getSql(),
+                                                                request.path(),
+                                                                format.getFormatName());
+            RestChannelConsumer result = newSqlQueryHandler.prepareRequest(newSqlRequest, client);
+            if (result != RestSQLQueryAction.NOT_SUPPORTED_YET) {
+                return result;
```
Comment (penghuo) [inline legacy/src/main/java/com/amazon/opendistroforelasticsearch/sql/legacy/plugin/RestSqlAction.java:154]: add some log to tracking the fallback behaviour for debug

Comment (dai-chen) [reply legacy/src/main/java/com/amazon/opendistroforelasticsearch/sql/legacy/plugin/RestSqlAction.java:154]: sure, will do!

File: core/src/main/java/com/amazon/opendistroforelasticsearch/sql/planner/logical/LogicalValues.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import com.amazon.opendistroforelasticsearch.sql.expression.LiteralExpression;
+import com.google.common.collect.ImmutableList;
+import java.util.List;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.ToString;
+
+/**
+ * Logical operator which is a sequence of literal list.
```
Comment (penghuo) [inline core/src/main/java/com/amazon/opendistroforelasticsearch/sql/planner/logical/LogicalValues.java:28]: Add some SQL VALUES related doc if possible.

Comment (dai-chen) [reply core/src/main/java/com/amazon/opendistroforelasticsearch/sql/planner/logical/LogicalValues.java:28]: Sure, adding more comments. Thanks!

**Pull request merged**